### PR TITLE
Add measure methods

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

This PR introduces the addition of the `measuredLine` methods within `Qgs(Multi)LineString`. It builds upon the code from PostGIS behind `ST_AddMeasure`. 

Expressions and processing will follow in subsequent PRs.

cc @benoitdm-oslandia @strk

### CODE REVIEW

File: src/core/geometry/qgslinestring.h [hunk 1]
```diff
@@ -1067,6 +1067,15 @@ class CORE_EXPORT QgsLineString: public QgsCurve
...
     QgsBox3D calculateBoundingBox3D() const override;
 
+
+    /**
+     * Re-write the measure ordinate (or add one, if it isn't already there) interpolating
+     * the measure between the supplied \a start and \a end values.
+     *
+     * \since QGIS 3.36
+     */
+    QgsLineString *measuredLine( double start, double end );
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.h:1077]: Missing SIP_FACTORY

File: src/core/geometry/qgslinestring.cpp [hunk 2]
```diff
@@ -2377,3 +2378,39 @@ void QgsLineString::transformVertices( const std::function<QgsPoint( const QgsPo
...
+  const int nbpoints = numPoints();
+  QgsLineString *cloned = clone();
+  if ( isEmpty() || ( nbpoints < 2 ) )
+  {
+    return cloned;
+  }
+
+  if ( !cloned->convertTo( QgsWkbTypes::addM( mWkbType ) ) )
+  {
+    qDebug() << "Impossible to convert type to M type\n";
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:2394]: Can you remove these or replace with QgsDebugError?

File: src/core/geometry/qgslinestring.cpp [hunk 3]
```diff
@@ -2377,3 +2378,39 @@ void QgsLineString::transformVertices( const std::function<QgsPoint( const QgsPo
...
+    return cloned;
+  }
+
+  const double range = end - start;
+  double line_length = length();
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:2399]: Should be camel case (elsewhere too)

File: src/core/geometry/qgslinestring.cpp [hunk 4]
```diff
@@ -2377,3 +2378,39 @@ void QgsLineString::transformVertices( const std::function<QgsPoint( const QgsPo
...
+  double length_so_far = 0.0;
+
+  cloned->setMAt( 0, start );
+  for ( int i = 1; i < nbpoints ; ++i )
+  {
+    length_so_far += QgsGeometryUtilsBase::distance2D( mX[ i - 1], mY[ i - 1 ], mX[ i ], mY[ i ] );
+    if ( line_length > 0.0 )
+      cloned->setMAt( i, start + range * length_so_far / line_length );
+    /* #3172, support (valid) zero-length inputs */
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:2408]: What's 3172 relate to?

File: src/core/geometry/qgsmultilinestring.cpp [hunk 5]
```diff
@@ -178,3 +179,40 @@ bool QgsMultiLineString::wktOmitChildType() const
...
 
+QgsMultiLineString *QgsMultiLineString::measuredLine( double start, double end )
+{
+
+
+  QgsMultiLineString *result{createEmptyWithSameType()};
+
+  if ( isEmpty() )
+  {
+    return result;
```
Comment (nyalldawson) [inline src/core/geometry/qgsmultilinestring.cpp:190]: Shouldn't this return an empty multilinstringM?

File: src/core/geometry/qgslinestring.cpp (hunk 4 above)
Comment (lbartoletti) [reply src/core/geometry/qgslinestring.cpp]: Copy/paste from a postgis comment.

**Review by benoitdm-oslandia — APPROVED** [review]
LGTM!

File: src/core/geometry/qgslinestring.cpp [hunk 6]
```diff
@@ -2377,3 +2378,38 @@ void QgsLineString::transformVertices( const std::function<QgsPoint( const QgsPo
   }
   clearCache();
 }
+
+
+QgsLineString *QgsLineString::measuredLine( double start, double end ) const
+{
+  const int nbpoints = numPoints();
+  QgsLineString *cloned = clone();
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:2386]:
```suggestion
  std::unique_ptr< QgsLineString > cloned( clone() );
```

File: src/core/geometry/qgslinestring.cpp [hunk 7]
```diff
@@ -2377,3 +2378,38 @@ void QgsLineString::transformVertices( const std::function<QgsPoint( const QgsPo
...
+
+  if ( !cloned->convertTo( QgsWkbTypes::addM( mWkbType ) ) )
+  {
+    return cloned;
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:2390]:
```suggestion
    return cloned.release();
```

File: src/core/geometry/qgslinestring.cpp [hunk 8]
```diff
@@ -2377,3 +2378,38 @@ void QgsLineString::transformVertices( const std::function<QgsPoint( const QgsPo
...
+  }
+
+  if ( isEmpty() || ( nbpoints < 2 ) )
+  {
+    return cloned;
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:2395]:
```suggestion
    return cloned.release();
```

File: src/core/geometry/qgslinestring.cpp [hunk 9]
```diff
@@ -2377,3 +2378,38 @@ void QgsLineString::transformVertices( const std::function<QgsPoint( const QgsPo
...
+  for ( int i = 1; i < nbpoints ; ++i )
+  {
+    lengthSoFar += QgsGeometryUtilsBase::distance2D( mX[ i - 1], mY[ i - 1 ], mX[ i ], mY[ i ] );
+    if ( lineLength > 0.0 )
+      cloned->setMAt( i, start + range * lengthSoFar / lineLength );
+    else if ( lineLength == 0.0 && nbpoints > 1 )
+      cloned->setMAt( i, start + range * i / ( nbpoints - 1 ) );
+    else
+      cloned->setMAt( i, 0.0 );
+  }
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp:2412]: Let's go for speed!

```suggestion
  
  double *mOut = cloned->mM.data();
  *mOut++ = start;
  for ( int i = 1; i < nbpoints ; ++i )
  {
    lengthSoFar += QgsGeometryUtilsBase::distance2D( mX[ i - 1], mY[ i - 1 ], mX[ i ], mY[ i ] );
    if ( lineLength > 0.0 )
      *mOut++ = start + range * lengthSoFar / lineLength;
    else if ( lineLength == 0.0 && nbpoints > 1 )
      *mOut++ = start + range * i / ( nbpoints - 1 );
    else
      *mOut++ = 0.0;
  }
```

File: src/core/geometry/qgsmultilinestring.cpp [hunk 10]
```diff
@@ -178,3 +179,39 @@ bool QgsMultiLineString::wktOmitChildType() const
...
+
+  if ( !result->convertTo( QgsWkbTypes::addM( mWkbType ) ) )
+  {
+    return result;
+  }
+
+  if ( isEmpty() )
+  {
+    return result;
+  }
```
Comment (nyalldawson) [inline src/core/geometry/qgsmultilinestring.cpp:196]: There's no benefit of doing things this way vs just creating a new QgsMultiLineString, as the first call to addGeometry for an empty multilinestring automatically sets the multilinestring z/m presence to match the first added geometry:

```suggestion
  std::unique_ptr< QgsMultiLineString > result = std::make_unique< QgsMultiLineString >();)
  if ( isEmpty() )
  {
    result->convertTo( QgsWkbTypes::addM( mWkbType ) )
    return result;
  }
```

File: src/core/geometry/qgsmultilinestring.cpp [hunk 11]
```diff
@@ -178,3 +179,39 @@ bool QgsMultiLineString::wktOmitChildType() const
...
+  const double range{end - start};
+  double lengthSoFar{0.0};
+
+  result->reserve( numGeometries() );
+  for ( int i = 0; i < numGeometries(); i++ )
+  {
+    const double subLength{geometryN( i )->length()};
+
+    const double subStart{ ( start + range *lengthSoFar / length ) };
+    const double subEnd{ ( start + range * ( lengthSoFar + subLength ) / length ) };
```
Comment (nyalldawson) [inline src/core/geometry/qgsmultilinestring.cpp:209]: Need some logic to handle length == 0 here to avoid divide by 0

Comment (lbartoletti) [reply src/core/geometry/qgsmultilinestring.cpp:202]: Will returns NaN as for PostGIS.

`SELECT ST_AsText(ST_AddMeasure('multilinestring(( 0 0, 0 0, 0 0), (1 1, 1 1, 1 1) )', 10 , 20)), ST_Length('multilinestring(( 0 0, 0 0, 0 0), (1 1, 1 1, 1 1) )')`

=> 
`MULTILINESTRING M ((0 0 NaN,0 0 NaN,0 0 NaN),(1 1 NaN,1 1 NaN,1 1 NaN)) | 0.0`

I added a test case

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**